File: studio/tis-versions/tis-3.6.0/ng-tis-3.6.0/src/runtime/misc/RCDeployment.ts
/**
 *   Licensed to the Apache Software Foundation (ASF) under one
 *   or more contributor license agreements.  See the NOTICE file
 *   distributed with this work for additional information
 *   regarding copyright ownership.  The ASF licenses this file
 *   to you under the Apache License, Version 2.0 (the
 *   "License"); you may not use this file except in compliance
 *   with the License.  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

import {BasicFormComponent} from "../../common/basic.form.component";
import {K8SRCSpec, K8SReplicsSpecComponent} from "../../common/k8s.replics.spec.component";
import FlinkJobDetail = flink.job.detail.FlinkJobDetail;
import {StepType} from "../../common/steps.component";
import {Descriptor} from "../../common/tis.plugin";

/**
 * Copyright (c) 2020 QingLang, Inc. <baisui@qlangtech.com>
 * <p>
 *   This program is free software: you can use, redistribute, and/or modify
 *   it under the terms of the GNU Affero General Public License, version 3
 *   or later ("AGPL"), as published by the Free Software Foundation.
 * <p>
 *  This program is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 *   FITNESS FOR A PARTICULAR PURPOSE.
 * <p>
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

export interface CpuLimit {
  unit: string;
  unitEmpty: boolean;
  val: number;
}

export interface CpuRequest {
  unit: string;
  unitEmpty: boolean;
  val: number;
}


export interface MemoryLimit {
  unit: string;
  unitEmpty: boolean;
  val: number;
}

export interface MemoryRequest {
  unit: string;
  unitEmpty: boolean;
  val: number;
}

export interface Status {
  availableReplicas: number;
  fullyLabeledReplicas: number;
  observedGeneration: number;
  readyReplicas: number;
  replicas: number;
}

export interface RCDeployment {
  cpuLimit: CpuLimit;
  cpuRequest: CpuRequest;
  creationTimestamp: number;
  dockerImage: string;
  envs: Map<string, string>;
  pods: Array<K8sPodState>;
  memoryLimit: MemoryLimit;
  memoryRequest: MemoryRequest;
  replicaCount: number;
  status: Status;
}

export interface K8sPodState {
  name: string;
  phase?: string;
  startTime?: string;
  restartCount?: number;
}

export enum LogType {
  INCR_DEPLOY_STATUS_CHANGE = "incrdeploy-change",
  DATAX_WORKER_POD_LOG = "datax-worker-pod-log"
}

export interface RcHpaStatus {
  conditions: Array<HpaConditionEvent>;
  currentMetrics: Array<HpaMetrics>;
  autoscalerStatus: HpaAutoscalerStatus;
  autoscalerSpec: HpaAutoscalerSpec;
}

export interface HpaConditionEvent {
  type: string;
  status: string;
  lastTransitionTime: string;
  reason: string;
  message: string;
}

export interface HpaAutoscalerStatus {
  currentCPUUtilizationPercentage: number;
  currentReplicas: number;
  desiredReplicas: number;
  lastScaleTime: number;
}

export interface HpaAutoscalerSpec {
  maxReplicas: number;
  minReplicas: number;
  targetCPUUtilizationPercentage: number;
}

export interface HpaMetrics {
  type: string;

  resource: UsingResource;
}

export interface UsingResource {
  name: string;
  currentAverageUtilization: any;
  currentAverageValue: any;
}

export class K8SControllerStatus {
  public k8sReplicationControllerCreated: boolean;

  public state: "NONE" | "STOPED" | "RUNNING" | "DISAPPEAR";
  // 由于本地执行器没有安装，导致datax执行器无法执行
  public installLocal: boolean;
  public rcDeployment: RCDeployment;
}

export interface PluginExtraProps {
  endType: string;
  extendPoint: string;
  supportIncr: boolean;
  impl: string;
  displayName: string;
}

export interface IncrDesc extends PluginExtraProps {
  extendSelectedTabProp: boolean;
}

export class IndexIncrStatus extends K8SControllerStatus {
  public incrScriptCreated: boolean;
  public incrScriptMainFileContent: string;
  public k8sPluginInitialized: boolean;
  public flinkJobDetail: FlinkJobDetail;
  public incrProcess: IncrProcess;

  public incrSourceDesc: IncrDesc;
  public incrSinkDesc: IncrDesc;
  public readerDesc: PluginExtraProps;
  public writerDesc: PluginExtraProps;


  public static getIncrStatusThenEnter(basicForm: BasicFormComponent, hander: ((r: IndexIncrStatus) => void), cache = true) {
    basicForm.httpPost('/coredefine/corenodemanage.ajax'
      , `action=core_action&emethod=get_incr_status&cache=${cache}`)
      .then((r) => {
        if (r.success) {
          let incrStatus: IndexIncrStatus = Object.assign(new IndexIncrStatus(), r.bizresult);
          hander(incrStatus);
        }
      });
  }

  /**
   * 增量处理节点启动有异常
   */
  public get incrProcessLaunchHasError(): boolean {
    return this.k8sReplicationControllerCreated && this.incrProcess && !this.incrProcess.incrGoingOn && !!this.rcDeployment && !!this.rcDeployment.status && this.rcDeployment.status.readyReplicas > 0;
  }

  public getFirstPod(): K8sPodState {
    for (let i = 0; i < this.rcDeployment.pods.length; i++) {
      return this.rcDeployment.pods[i];
      break;
    }
    return null;
  }
}

export interface IncrProcess {
  incrGoingOn: boolean;
  incrProcessPaused: boolean;
}

export interface ProcessMeta {
  targetName: string;
  pageHeader: string;
  stepsType: StepType
  supportK8SReplicsSpecSetter: boolean;
  // step0
  notCreateTips: string;
  createButtonLabel: string;
}

export class DataXJobWorkerStatus extends K8SControllerStatus {
  processMeta: ProcessMeta;
}

export class DataxWorkerDTO {
  rcSpec: K8SRCSpec;
  processMeta: ProcessMeta;
}

File: market/tis-versions/tis-3.6.0/tis/tis-builder-api/src/main/java/com/qlangtech/tis/coredefine/module/action/impl/RcDeployment.java
/**
 *   Licensed to the Apache Software Foundation (ASF) under one
 *   or more contributor license agreements.  See the NOTICE file
 *   distributed with this work for additional information
 *   regarding copyright ownership.  The ASF licenses this file
 *   to you under the Apache License, Version 2.0 (the
 *   "License"); you may not use this file except in compliance
 *   with the License.  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
package com.qlangtech.tis.coredefine.module.action.impl;

import com.qlangtech.tis.config.k8s.ReplicasSpec;
import com.qlangtech.tis.coredefine.module.action.IDeploymentDetail;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 运行时RC
 *
 * @author 百岁（baisui@qlangtech.com）
 * @create: 2020-09-02 12:32
 */
public class RcDeployment extends ReplicasSpec implements IDeploymentDetail {

    // 创建时间
    private long creationTimestamp;

    private ReplicationControllerStatus status;

    private List<PodStatus> pods = new ArrayList<>();

    // 环境变量
    final Map<String, String> envs = new HashMap<>();

    private String dockerImage;

    public ReplicationControllerStatus getStatus() {
        return status;
    }

    public void setStatus(ReplicationControllerStatus status) {
        this.status = status;
    }

    public String getDockerImage() {
        return dockerImage;
    }

    public void addPod(PodStatus podStat) {
        this.pods.add(podStat);
    }

    public List<PodStatus> getPods() {
        return pods;
    }

    public void setDockerImage(String dockerImage) {
        this.dockerImage = dockerImage;
    }

    public void addEnv(String key, String val) {
        this.envs.put(key, val);
    }

    public Map<String, String> getEnvs() {
        return this.envs;
    }

    public long getCreationTimestamp() {
        return creationTimestamp;
    }

    public void setCreationTimestamp(long creationTimestamp) {
        this.creationTimestamp = creationTimestamp;
    }

    @Override
    public void accept(IDeploymentDetailVisitor visitor) {
        visitor.visit(this);
    }

    public static class ReplicationControllerStatus {

        private Integer availableReplicas;

        private Integer fullyLabeledReplicas;

        private Long observedGeneration;

        // 已经有的副本
        private Integer readyReplicas;

        // 目标副本
        private Integer replicas;

        public Integer getAvailableReplicas() {
            return availableReplicas;
        }

        public void setAvailableReplicas(Integer availableReplicas) {
            this.availableReplicas = availableReplicas;
        }

        public Integer getFullyLabeledReplicas() {
            return fullyLabeledReplicas;
        }

        public void setFullyLabeledReplicas(Integer fullyLabeledReplicas) {
            this.fullyLabeledReplicas = fullyLabeledReplicas;
        }

        public Long getObservedGeneration() {
            return observedGeneration;
        }

        public void setObservedGeneration(Long observedGeneration) {
            this.observedGeneration = observedGeneration;
        }

        public Integer getReadyReplicas() {
            return readyReplicas;
        }

        public void setReadyReplicas(Integer readyReplicas) {
            this.readyReplicas = readyReplicas;
        }

        public Integer getReplicas() {
            return replicas;
        }

        public void setReplicas(Integer replicas) {
            this.replicas = replicas;
        }
    }

    public static class PodStatus {

        private String name;

        private String phase;

        // pod启动次数
        private int restartCount;

        private long startTime;

        public String getName() {
            return name;
        }

        public int getRestartCount() {
            return restartCount;
        }

        public void setRestartCount(int restartCount) {
            this.restartCount = restartCount;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getPhase() {
            return phase;
        }

        public void setPhase(String phase) {
            this.phase = phase;
        }

        public long getStartTime() {
            return startTime;
        }

        public void setStartTime(long startTime) {
            this.startTime = startTime;
        }
    }
}
